vllm/v1/sample/rejection_sampler.py

```diff
@@ -318,18 +366,39 @@ def expand_batch_to_tokens(
     Returns:
         expanded_x: [num_tokens] tensor.
     """
-    batch_size = x.shape[0]
-    assert cu_num_tokens.shape[0] == batch_size
-    expanded_x = x.new_empty(num_tokens)
-    expand_kernel[(batch_size, )](
-        expanded_x,
-        x,
-        cu_num_tokens,
-        replace_from,
-        replace_to,
-        MAX_NUM_TOKENS=MAX_SPEC_LEN,  # To avoid recompilation.
-        num_warps=1,
-    )
+    if x.ndim == 1:
+        batch_size = x.shape[0]
+        assert cu_num_tokens.shape[0] == batch_size
+        expanded_x = x.new_empty(num_tokens)
+
+        expand_kernel[(batch_size, )](
+            expanded_x,
+            x,
+            cu_num_tokens,
+            replace_from,
+            replace_to,
+            MAX_NUM_TOKENS=MAX_SPEC_LEN,  # To avoid recompilation.
+            num_warps=1,
+        )
+    elif x.ndim == 2:
+        batch_size, vocab_size = x.shape
+        assert cu_num_tokens.shape[0] == batch_size
+        expanded_x = x.new_empty((num_tokens, vocab_size))
+
+        expand_kernel_2d[(batch_size, vocab_size)](
+            expanded_x,
+            x,
+            cu_num_tokens,
+            replace_from,
+            replace_to,
+            MAX_NUM_TOKENS=MAX_SPEC_LEN,
+            VOCAB_SIZE=vocab_size,
+            num_warps=4 if vocab_size > 512 else 1,
+        )
```

Comment:
If I understand correctly, this kernel is not vectorized enough, and therefore uses too many thread blocks. Please fix the kernel to do vectorized load. If it is implemented correctly, we should launch either `(batch_size,)` or `(batch_size, cdiv(vocab_size, BLOCK_SIZE)` thread blocks (depending on how it is implemented).